flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -281,10 +289,41 @@ protected void processInput(DefaultActionContext context) throws Exception {
 		if (status == InputStatus.END_OF_INPUT) {
 			context.allActionsCompleted();
 		}
-		else if (status == InputStatus.NOTHING_AVAILABLE) {
+
+		CompletableFuture<?> jointFuture = getInputOutputJointFuture(status);
+		if (jointFuture != null) {
 			SuspendedMailboxDefaultAction suspendedDefaultAction = context.suspendDefaultAction();
-			inputProcessor.getAvailableFuture().thenRun(suspendedDefaultAction::resume);
+			jointFuture.thenRun(suspendedDefaultAction::resume);
+		}
+	}
+
+	/**
+	 * @return a combination of input and output futures if at-least one future of them is not
+	 * completed, otherwise return null if all input and outputs are available.
+	 */
+	private CompletableFuture<?> getInputOutputJointFuture(InputStatus status) {
+		if (status == InputStatus.MORE_AVAILABLE && isOutputAvailable()) {
+			return null;
+		}
+
+		int length = recordWriters.size();
+		for (int i = 0; i < length; i++) {
+			inputOutputFutures[i] = recordWriters.get(i).getAvailableFuture();
+		}
+		inputOutputFutures[length] = inputProcessor.getAvailableFuture();
+		return CompletableFuture.allOf(inputOutputFutures);
```

Comment:
Shouldn't we distinguish three separate cases:

1. just input unavailable
2. just output unavailable
3. both unavailable

and return only a relevant future for each of the case? It might be faster slightly faster and a bit cleaner (assuming that the logic of doing so won't be too complicated, but it shouldn't)
```
boolean outputUnavailable = !recordWriters.isAvailable();
if (inputStatus == NOTHING_AVAILABLE && outputUnavailable) 
  return CompletableFuture.allOf(inputProcessor.getAvailableFuture(), recordWriters.getAvailableFuture());
else if (inputStatus == NOTHING_AVAILABLE) 
  return inputProcessor.getAvailableFuture();
else if (outputUnavailable)
  return recordWriters.getAvailableFuture();
```
both inputs and outputs being unavailable will be a veeerryy rare temporary case (having both nothing to process and a back-pressure).

Reply:
I agree it would be a bit clearly to distinguish three cases, and my previous way was considering these three cases together via the condition of `if (status == InputStatus.MORE_AVAILABLE && isOutputAvailable())` to make it shorter.  In other words, as long as at-least one input/output is not available, all the futures would be squashed together(including some completed future).  

Another concern to do so is that even though we distinguish the input with output, but we can not further distinguish among outputs. I mean even though only one `RecordWriter` is not available, but we still need to compose all the other completed `RecordWriter` futures to avoid constructing the array repeatedly.

I would take your way to best-effort exclude non-relevant futures.